### PR TITLE
ApplicativeError

### PR DESCRIPTION
**Pull request by GrafBlutwurst**
A very minimal implementation just by seperating out stuff from `MonadError` 

if there are sensible functions to put on `ApplicativeError` let me know!

### CODE REVIEW

File: core/src/main/scala/scalaz/MonadError.scala [hunk 1]
```diff
@@ -5,18 +5,13 @@ package scalaz
...
+trait MonadError[F[_], S] extends Monad[F] with ApplicativeError[F, S] { self =>
   ////
 
-  def raiseError[A](e: S): F[A]
-  def handleError[A](fa: F[A])(f: S => F[A]): F[A]
-
   def emap[A, B](fa: F[A])(f: A => S \/ B): F[B] =
     bind(fa)(a => f(a).fold(raiseError(_), pure(_)))
 
   trait MonadErrorLaw {
```
Comment (hrhino) [inline core/src/main/scala/scalaz/MonadError.scala]: should extend `ApplicativeErrorLaw`, no?

File: core/src/main/scala/scalaz/ApplicativeError.scala [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package scalaz
+
+trait ApplicativeError[F[_], S] extends Applicative[F]{
+
+  ////
```
Comment (hrhino) [inline core/src/main/scala/scalaz/ApplicativeError.scala]: the point of the slashes is that you should modify `GenTypeClass` to generate this file (and `ApplicativeSyntax`) for you with a standard template; see #2004 for an example

File: core/src/main/scala/scalaz/ApplicativeError.scala [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  trait ApplicativeErrorLaws{
+    def raisedErrorsHandled[A](e: S, f: S => F[A])(implicit FEA: Equal[F[A]]): Boolean =
+      FEA.equal(handleError(raiseError(e))(f), f(e))
+  }
+
+  def applicativeErrorLaws = new ApplicativeErrorLaws {}
+
+  def raiseError[A](e:S):F[A]
+  def handleError[A](fa: F[A])(f: S => F[A]): F[A]
```
Comment (hrhino) [inline core/src/main/scala/scalaz/ApplicativeError.scala]: Please move the abstract members to the top of the template, so it's easier to see what needs to be implemented.

File: core/src/main/scala/scalaz/Validation.scala [hunk 4]
```diff
@@ -545,8 +545,8 @@ sealed abstract class ValidationInstances3 {
         fab.bitraverse(f, g)
     }
 
-  implicit def ValidationApplicative[L: Semigroup]: Applicative[Validation[L, ?]] with Alt[Validation[L, ?]] =
-    new Applicative[Validation[L, ?]] with Alt[Validation[L, ?]] {
+  implicit def ValidationApplicativeError[L: Semigroup]: ApplicativeError[Validation[L, ?], L] with Alt[Validation[L, ?]] =
```
Comment (hrhino) [inline core/src/main/scala/scalaz/Validation.scala]: 💯

File: tests/src/test/scala/scalaz/DisjunctionTest.scala [hunk 5]
```diff
@@ -11,6 +11,7 @@ object DisjunctionTest extends SpecLite {
   checkAll(monoid.laws[Int \/ Int])
   checkAll(bindRec.laws[Int \/ ?])
   checkAll(monad.laws[Int \/ ?])
+  checkAll(applicativeError.laws[Int \/ ?, Int])
```
Comment (hrhino) [inline tests/src/test/scala/scalaz/DisjunctionTest.scala]: if you make `MonadErrorLaws extends ApplicativeErrorLaws` you don't need this, right?

File: core/src/main/scala/scalaz/syntax/ApplicativeErrorSyntax.scala [hunk 6]
```diff
@@ -0,0 +1,30 @@
+package scalaz
+package syntax
+
+/** Wraps a value `self` and provides methods related to `ApplicativeError` */
+final class ApplicativeErrorOps[F[_], S, A] private[syntax](self: F[A])(implicit val F: ApplicativeError[F, S]) {
+  ////
+
```
Comment (xuwei-k) [inline core/src/main/scala/scalaz/syntax/ApplicativeErrorSyntax.scala]: https://github.com/scalaz/scalaz/blob/e8ae7157a89a9fdf020fa65f7e6d0c396d77075e/core/src/main/scala/scalaz/syntax/MonadErrorSyntax.scala

We should move some methods from `MonadError`.

File: scalacheck-binding/src/main/scala/scalaz/scalacheck/ScalazProperties.scala [hunk 7]
```diff
@@ -686,7 +695,6 @@ object ScalazProperties {
     def laws[F[_], E](implicit me: MonadError[F, E], am: Arbitrary[F[Int]], afap: Arbitrary[F[Int => Int]], aeq: Equal[F[Int]], ae: Arbitrary[E], afea: Arbitrary[E => F[Int]]): Properties =
       newProperties("monad error"){ p =>
         p.include(monad.laws[F])
```
Comment (xuwei-k) [inline scalacheck-binding/src/main/scala/scalaz/scalacheck/ScalazProperties.scala]:
```diff
  p.include(monad.laws[F])
+ p.include(applicativeError.laws[F])
- p.property("raisedErrorsHandled") = raisedErrorsHandled[F, E, Int]
```

Comment (GrafBlutwurst) [reply scalacheck-binding/src/main/scala/scalaz/scalacheck/ScalazProperties.scala]: I can't figure out where scalacheck-binding/src/main/scala/scalaz/scalacheck/ScalazProperties.scala is generated. can you please point me to it?

**Comment by xuwei-k** [comment]
Thanks!

**Pull request merged**